dotcom-rendering/src/model/insert-newsletter-signup.ts

```diff
@@ -0,0 +1,167 @@
+interface PlaceInArticle {
+	index: number;
+	distance: number;
+	afterText: boolean;
+	afterSupporting: boolean;
+}
+
+const isCloseAfterSupportingElement = (
+	index: number,
+	elements: CAPIElement[],
+): boolean => {
+	const elementIsSupporting = (element?: CAPIElement): boolean =>
+		!!(element && 'role' in element && element.role === 'supporting');
+
+	return (
+		elementIsSupporting(elements[index - 1]) ||
+		elementIsSupporting(elements[index - 2])
+	);
+};
+
+const isAfterText = (index: number, elements: CAPIElement[]): boolean =>
+	elements[index - 1]?._type ===
+	'model.dotcomrendering.pageElements.TextBlockElement';
+
+const getPlaces = (
+	targetIndex: number,
+	elements: CAPIElement[],
+): PlaceInArticle[] => {
+	return elements.map((element, index) => ({
+		index,
+		distance: Math.abs(index - targetIndex),
+		afterText: isAfterText(index, elements),
+		afterSupporting: isCloseAfterSupportingElement(index, elements),
+	}));
+};
+
+const placeIsSuitable = (place: PlaceInArticle): boolean =>
+	place.afterText && !place.afterSupporting;
+
+/**
+ * Finds a place within the existing element of the block to insert a
+ * NewsletterSignupBlockElement, within a given distance from a given
+ * target index, prefering a place after TextBlockElement and not within
+ * two places after a supporting element (eg pullquote).
+ *
+ * @param elements the elements in the article block
+ * @param targetIndex the target place to put the NewsletterSignupBlockElement
+ * @param maxDistance how far to move from the targetIndex to find a place after a paragraph
+ * @returns the index insert the NewsletterSignupBlockElement component at
+ */
+export const findInsertIndex = (
+	elements: CAPIElement[],
+	targetIndex: number,
+	maxDistance = 0,
+): number => {
+	const possiblePlaces = getPlaces(targetIndex, elements)
+		.filter(placeIsSuitable)
+		.filter((place) => place.distance <= maxDistance)
+		.sort((a, b) => a.distance - b.distance); // sort, closest place first
+
+	// Return index of the closest place - if there are none, return the targetIndex
+	return possiblePlaces[0]?.index || targetIndex;
+};
+
+/**
+ * Determines whether the element should have clear:both
+ * for all breakpoints
+ *
+ * @param index where in the array the element will be inserted
+ * @param elements the array of elements
+ * @returns whether the element should have clear:both for all breakpoints
+ */
+const decideIfClearFromLeftCol = (
```

Comment:
This is a little clearer after the changes for the previous comment - it's not testing if the *immediately preceding* element is has the role 'supporting', 'thumbnail' or 'richLink', but finding the last element with one of those roles that comes before the point the `EmailSignUp` is inserted. That context is used to change the css on the `EmailSignUp`.

Being close after  a 'thumbail' or 'richlink' doesn't actually make a place unsuitable as they just float the the left column- its only the 'supporting' element that are problematic. The suitability logic does take account of this and tries to place the EmailSignUp at least two places after any 'supporting' element.

I think we should work on the assumption that the EmailSignUp might end up in a "not ideal" position but make sure that the css adapts to the context:

When the last floating element is supporting, we need to set `clear:both` to avoid this scenario:
<img width="400" alt="Screenshot 2022-07-18 at 14 32 06" src="https://user-images.githubusercontent.com/30567854/181213370-3b7096b0-a600-41c3-971a-d97ef165aa60.png">

But we can't have `clear:both` as standard or this will happen near a 'thumbail' or 'richlink' or we get this:

<img width="400" alt="thumbBad" src="https://user-images.githubusercontent.com/30567854/181218899-7ab5bad0-b8ec-4e0f-9eed-3fd91725ead3.png">

Reply:
The css logic looks at "last floating element before" rather than just the immediately preceding because the floating block can be taller than the blocks between it and sign up block - the placement logic can't account for that, so we need CSS to avoid spacing/layout issue

Reply:
On reflection, I think to start with we should prefer not inserting the embed rather than adding this complexity. Article elements in DCR are self contained and only vary their look or behaviour based on `format` and this is a contract that we want to preserve.

If, at a later date, we identify a problem where embeds are being lost because of this problem then we can revisit this decision but at this stage I think this is an edge case and it is better to either place the embed somewhere else or not at all.

We should be leaning towards *only* inserting when "ideal", rather than allowing 'not ideal' positions and then having to add code to manage this.

Reply:
One issue there is that it's not possible to determine with certainty whether a position is truely 'ideal' in relation to previous floating elements based on the array of elements and their types. Whether a floating elements will overlap with the sign-up block depends on the relative heights of the floating element and the elements between it and the sign-up block, which is determined by the styling as well as the content. 

This is an extreme example to illustrate, but this position of the sign-up block would be considered "ideal" from the placement logic as it has five paragraphs between it and the 'supporting' block, but since the floating block is taller than those five paragraphs, it will still interfere with the sign-up block if we don't give the sign-up block the context telling it how to style itself  (with `clear:both`) :
<img width="400" alt="Screenshot 2022-07-27 at 15 01 12" src="https://user-images.githubusercontent.com/30567854/181267488-ddab03ee-fa9a-48f2-86ef-34b19378c578.png">

Reply:
If can't provide the context it would be best to  always use `clear:both` - we could then get the equivalent issue with the 'thumbnail' , but that's  probably less bad (again, apologies for the extreme example): 
<img width="400" alt="Screenshot 2022-07-27 at 15 11 49" src="https://user-images.githubusercontent.com/30567854/181270010-9c1c91f7-09b9-4cc9-b815-a8af43218324.png">

Reply:
Aside from that, I think there's a good case for default to the end of the article if we can't find a place near the middle that is (as far as we can tell) "ideal". Let me consult with our PM on that.

Reply:
Pushed some changes - the `EmailSignUp` component no longer gets the prop telling it what the last floating element was, so it is self contained again. It will always have `clear:both` on the container css.

To minimise the 'blank space' problem illustrated above, I've :
 - made the `isSuitable` test more strict - the position now has to be at least 4 spaces after *any* floating element (rather than at least 2 after a supporting as before)
 - changed the sorting logic so the distance after a floating element is more important than the distance from the target (IE we prefer places further away from the floating element over places nearer the target)
 - defaults to placeing the `EmailSignUp` at the **end** (not the `targetIndex`) of the article if no suitable place can be found

Reply:
Editorial want to be able as assess the impact of the sign up block being put at the end of the article, so I've added a logging statement to track when this happens - the concerns are that it could be less effective and might overshadow (of be overshadowed by) the contributions asks at the end of the page.